distribution/tools/plugin-cli/src/test/java/org/elasticsearch/plugins/cli/InstallPluginActionTests.java

```diff
@@ -1507,4 +1545,44 @@ public void testInstallMigratedPlugins() throws Exception {
             assertThat(terminal.getErrorOutput(), containsString("[" + id + "] is no longer a plugin"));
         }
     }
+
+    public void testStablePluginWithNamedComponentsFile() throws Exception {
+        InstallablePlugin stablePluginZip = createStablePlugin("stable1", pluginDir, true);
+        installPlugins(List.of(stablePluginZip), env.v1());
+        assertPlugin("stable1", pluginDir, env.v2());
+        assertNamedComponentFile("stable1", env.v2().pluginsFile(), namedComponentsJSON());
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testStablePluginWithoutNamedComponentsFile() throws Exception {
+        // named component will have to be generated upon install
+        InstallablePlugin stablePluginZip = createStablePlugin("stable1", pluginDir, false);
+
+        installPlugins(List.of(stablePluginZip), env.v1());
+
+        assertPlugin("stable1", pluginDir, env.v2());
+        // the test was using a jar with no classes, so the scanner did not find anything, but this is fine. that logic is tested elsewhere
```

Comment:
I think we care about slightly more than that. Is there a legitimate case for an empty file? I think we want to see that the generation was called and a real file was created, with at least one actual component. I would do that in writeStablePlugin using the InMemoryJavaCompiler.

Reply:
fair point, InMemoryJavaCompiler might help with test here and it helped me realise that plugin-cli has to have a dependency on plugin-api. I incorrectly assumed that it might be shipped with plugin, but it will not. Plugins only have 'implementation/compileOnly' scope on it.